jbpm-runtime-manager/src/main/java/org/jbpm/runtime/manager/impl/lock/RuntimeManagerLockWatcher.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jbpm.runtime.manager.impl.lock;
+
+import java.util.PriorityQueue;
+import java.util.Queue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import org.jbpm.runtime.manager.spi.RuntimeManagerLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class RuntimeManagerLockWatcher {
+    protected static final Logger logger = LoggerFactory.getLogger(RuntimeManagerLockWatcher.class);
+
+    private Long watchLockPolling;
+    private Long maxLockProcessingTime;
+
+    private ExecutorService executorService;
+    private Queue<LockWatchInfo> locksWatched;
+
+    class RuntimeManagerLockReaper implements Callable<Void> {
+
+        @Override
+        public Void call() throws Exception {
+            while(!Thread.currentThread().isInterrupted()) {
+                long maxTimeToWait = reap();
+                Thread.sleep(maxTimeToWait);
+            }
+            return null;
+        }
+
+    }
+
+    public long reap() {
+        synchronized (locksWatched) {
+            while(!locksWatched.isEmpty()) {
+                LockWatchInfo info = locksWatched.peek();
+                long idleTime = info.getCurrentProcessingTime();
+                if(maxLockProcessingTime > idleTime) {
+                    return maxLockProcessingTime - idleTime;
+                }
+                try {
+                    long timeExceeded = idleTime - maxLockProcessingTime;
+                    logger.info("Max process time lock exceeded for process instance id {} by {} ms. Trying forceful release lock {}", info.processInstanceId, timeExceeded, info.lock);
+                    info.lock.forceUnlock();
+                } catch(UnsupportedOperationException e) {
+                    logger.warn("Runtime manager lock implementation {} does not allow self release for process instance id {}", info.lock.getClass().getName(), info.processInstanceId);
+                }
+                // stop watching this lock. We gave already the information
+                locksWatched.poll();
```

Comment:
The problem with that is that if the lock does not support  forceUnlock it won't remove it. We want just one stacktrace, otherwise it will be flooding the logs.
wdyt ?